File: restif/Restif.py
from http.server import * 
import mysql.connector
import sys
import ast
import base64
class RestifHandler(BaseHTTPRequestHandler):
    def Connect(self,db):
        usr = 'admin'
        pwd = 'apwd'
        h = self.headers['Authorization']
        if h is not None:
            d = str(base64.b64decode(h[6:len(h)]),'utf-8') # 6 is len('Basic ')
            s = d.split(':')
            usr = s[0]
            pwd = s[1]
        return mysql.connector.connect(user=usr,password=pwd,database=db)
    def Send(self,status,mess):
        self.send_response(status)
        self.send_header("Cache-control", "no-store");
        self.send_header("Expires", "-1");
        self.send_header("Pragma", "no-cache");
        if mess!=None:
            bs = bytes(mess,'utf-8')
            self.send_header('Content-Length',len(bs))
            self.end_headers()
            self.wfile.write(bs) 
        return
    def do_GET(self):
        try:
            segments = self.path.split('/') 
            db = None
            tb = None
            wh = None
            if segments[1]!='':
                db = segments[1]
            if len(segments)>2:
                tb = segments[2]
            if len(segments)>3:
                wh = segments[3]
            conn = self.Connect(db)
            if conn is None:
                self.Send(400,'Cannot connect')
                return
            if tb==None:
                mess = self.Get1(conn,db)
            else:
                mess = self.Get2(conn,tb,wh)
            self.Send(200,mess)
        except Exception as e:
            self.Send(403,e.msg)
        if conn!=None:
            conn.close()
        return
    def Get1(self,conn,db):
        cursor = conn.cursor(raw=True)
        if db is None:
           cursor.execute('show databases')
        else:
            cursor.execute('show tables')
        mess = str()
        for rw in cursor.fetchall():
            mess += rw[0].decode('utf-8') + '\n'
        cursor.close()
        return mess
    def Get2(self,conn,tb,wh):
        cursor = conn.cursor()
        if wh is None:
            cursor.execute('Select * from '+tb)
        else:
            cursor.execute('Select * from '+tb+' where '+wh)
        rows = cursor.fetchall()
        desc = cursor._description
        acc = self.headers['Accept']
        fmts = { 'text/plain':['','\n','','',',',''], # array start mid end; obj start mid end
                'application/json':['[',',\n',']','{',', ','}']}
        fmt = fmts[acc]
        mess = []
        mess.append(fmt[0]) # array start
        rsep = ''     # row separator initially empty
        for rw in rows:
            mess.append(rsep) 
            rsep = fmt[1] # row sep is array mid
            mess.append(fmt[3]) # obj start
            fsep = ''
            for i in range(len(desc)):
                mess.append(fsep)
                fsep = fmt[4] # obj mid
                mess.append(self.Field(acc,desc,rw,i)) 
            mess.append(fmt[5]) # obj end
        cursor.close()
        mess.append(fmt[2]) # array end
        return ''.join(mess)
    def Field(self,acc,desc,data,i):
        v = self.Value(desc[i][1],str(data[i]))
        if acc=='text/plain':
            return v
        return '"' + desc[i][0] +'": ' + v
    def Value(self,typ,val):
        q = ""
        if typ in {253,254,252,10,7,11}: # char, varchar, text, date, timestamp, time
            q = "'" # single quote because Sql option uses text/plain too
        return q+val+q
    def log_request(code,size):
        return
    def GetData(self):
        if not self.headers.__contains__('Content-Length'):
            return None
        n = int(self.headers['Content-Length'])
        return str(self.rfile.read(n),'utf-8')
    def do_POST(self):
        segments = self.path.split('/') 
        db = None
        tb = None
        if segments[1]!='':
            db = segments[1]
        if len(segments)>2:
            tb = segments[2]
        data = self.GetData()
        if data is None:
            self.Send(400,'Nothing to do')
            return
        conn = self.Connect(db)
        if conn is None:
            self.Send(400,'Cannot connect')
            return
        if tb is None:
            self.Post1(conn,data)
        else:
            self.Post2(conn,tb,data)
        conn.close()
        return
    def Post1(self,conn,data):
        try:    
            conn.start_transaction(consistent_snapshot=True,isolation_level='SERIALIZABLE')
            k = 1
            for result in conn.cmd_query_iter(data):
                k = k+1
            conn.commit()
            self.Send(200,'OK')
        except Exception as e:
            self.Send(403,'line '+str(k)+': '+e.msg)
        return
    def Post2(self,conn,tb,data):
        try:
            data = ast.literal_eval(data)
            cmd = ['insert into '+tb+' (']
            vals = [') values (']
            cm = '' 
            for k in data.keys():
                cmd.append(cm+k)
                v = data[k]
                q = ''
                if type(v) is str:
                    q = "'"
                vals.append(cm+q+str(v)+q)
                cm = ','
            conn.cmd_query(''.join(cmd)+''.join(vals)+')')
            conn.commit()
            self.Send(200,'OK')
        except Exception as e:
            self.Send(403,e.msg)
        return
    def do_PUT(self):
        segments = self.path.split('/') 
        db = None
        tb = None
        wh = None
        if segments[1]!='':
            db = segments[1]
        if len(segments)>2:
            tb = segments[2]
        if len(segments)>>3:
            wh = segments[3]
        data = self.GetData()
        if data is None:
            self.Send(400,'Nothing to do')
            return
        conn = self.Connect(db)
        if conn is None:
            self.Send(400,'Cannot connect')
            return
        try:
            data = ast.literal_eval(data)
            cmd = ['update '+tb+' set ']
            cm = '' 
            for k in data.keys():
                v = data[k]
                q = ''
                if type(v) is str:
                    q = "'"
                cmd.append(cm+k+'='+q+str(v)+q)
                cm = ','
            conn.cmd_query(''.join(cmd)+' where '+wh)
            conn.commit()
            self.Send(200,'OK')
        except Exception as e:
            self.Send(403,e.msg)
        conn.close()
        return
    def do_DELETE(self):
        segments = self.path.split('/') 
        db = None
        tb = None
        wh = None
        if segments[1]!='':
            db = segments[1]
        if len(segments)>2:
            tb = segments[2]
        if len(segments)>>3:
            wh = segments[3]
        conn = self.Connect(db)
        if conn is None:
            self.Send(400,'Cannot connect')
            return
        try:
            data = ast.literal_eval(data)
            cmd = 'delete from '+tb
            if wh is not None:
                cmd += ' where '+wh
            conn.cmd_query(cmd)
            conn.commit()
            self.Send(200,'OK')
        except Exception as e:
            self.Send(403,e.msg)
        conn.close()
        return
try:
    server = HTTPServer(('',8088),RestifHandler)
    server.serve_forever()
except KeyboardInterrupt:
    print ('Exiting')
    server.socket.close()

File: restif/RestifD/RestifD/Restif.cs
using System;
using System.Text;
using System.Net;
using System.Collections.Generic;
using System.Threading.Tasks;
using MySql.Data.MySqlClient;
using System.IO;
namespace RestifD
{
    class Restif
    {
        HttpListenerContext context;
        static string host = "localhost";
        MySqlConnection conn = null;
        string origin = null;
        static void Main(string[] args)
        {
            if (args.Length > 0)
                host = args[0];
            var listener = new HttpListener();
            listener.Prefixes.Add("http://" + host + ":8078/");
            listener.Prefixes.Add("https://"+host+":4437/");
            listener.Realm = "MySQL server on " + host;
            listener.AuthenticationSchemes = AuthenticationSchemes.Basic;
            listener.Start();
            for (;;)
                try
                {
                    new Restif { context = listener.GetContext() }.Run();
                }
                catch(Exception e)
                {
                    Console.WriteLine(e.Message);
                    break;
                }
        }
        string get(string[] p, int i)
        {
            return (i < p.Length) ? (p[i] != "") ? p[i].Trim('/') : null : null;
        }
        void Run()
        {
            Task.Run(() =>
             {
                var meth = context.Request.HttpMethod;
                var path = context.Request.Url.Segments;
                origin = context.Request.Headers["Origin"];
                var h = context.Request.Headers["Authorization"];
                var s = Encoding.UTF8.GetString(Convert.FromBase64String(h.Substring(6))).Split(':');
                string db = get(path, 1);
                string tb = get(path, 2);
                string wh = get(path, 3);
                var page = db != null && (db.EndsWith(".htm")||db.EndsWith(".js"));
                var connstring = "server=" + host + ";uid="+s[0]+";password="+s[1];
                if (db != null)
                    connstring += ";database=" + db;
                string mess = null;
                try
                {
                    if (!page)
                        conn = new MySqlConnection(connstring);
                    conn?.Open();
                    switch (meth)
                    {
                        case "OPTIONS":
                           SendCORS();
                           return;
                        case "GET":
                          if (page)
                          {
                              var rdr = new StreamReader(db);
                              mess = rdr.ReadToEnd();
                              rdr.Close();
                          }
                          else
                              mess = (tb == null) ? Get1(db) : Get2(tb, wh);
                            break;
                        case "POST":
                            var pd = Receive();
                            if (tb != null)
                                mess = Post2(tb, new DocArray(pd));
                            else if (db != null)
                                mess = Post1(pd);
                            else
                                throw new Exception("not supported");
                            break;
                        case "PUT":
                            if (db == null || tb == null || wh == null)
                                throw new Exception("not supported");
                            mess = Put(tb,wh,new Document(Receive()));
                            break;
                        case "DELETE":
                            if (db == null || tb == null)
                                throw new Exception("not supported");
                            mess = Delete(tb,wh);
                            break;
                    }
                    Send(200, mess);
                } catch (Exception e)
                {
                    Send(403, e.Message);
                     Console.WriteLine("--> "+e.Message);
                }
                conn?.Close();
            });
        }
        string Get1(string db)
        {
            var cmd = conn.CreateCommand();
            cmd.CommandText = (db==null)?"show databases":"show tables";
            var rdr = cmd.ExecuteReader();
            var r = "";
            while (rdr.Read())
                r += " " + rdr[0].ToString();
            return r;
        }
        string Get2(string tb, string wh)
        {
            var cmd = conn.CreateCommand();
            cmd.CommandText = "select * from " + tb;
            if (wh != null)
                cmd.CommandText += " where " + wh;
            var rdr = cmd.ExecuteReader();
             var acc = context.Request.Headers["Accept"];
            //  0: fieldstart 1: fieldend 2: liststart 3: listmid 4: listend 
            //  5: rowstart 6: rowmid 7: rowend  
            var fmt = new string[] { "", "=", "", "\n", "", "", ",", ";" };
            switch (acc)
            {
                case "application/json":
                    fmt = new string[] { "\"", "\": ", "[", ",\n", "]", "{", ", ", "}" };
                    break;
                case "text/html":
                    fmt = new string[] { "", "", "<table>", "", "</table>", "<tr>","", "</tr>" };
                    break;
                case "application/xml":
                    fmt = new string[] { "<", ">", "<root>", "", "</root>", "<" + tb + ">", "", "</" + tb + ">" };
                    break;
            }
            var r = new StringBuilder(fmt[2]);
            var pos = 0;
            var m = "";
            while (rdr.Read())
            {
                if (acc=="text/html" && pos++ ==0) // generate headers row
                {
                    r.Append("<tr>");
                    for (var i = 0; i < rdr.VisibleFieldCount; i++)
                        r.Append("<th>" + rdr.GetName(i) + "</th>");
                    r.Append("</tr>");
                }
                var c = "";
                r.Append(m);
                m = fmt[3];
                r.Append(fmt[5]);
                for (var i = 0; i < rdr.VisibleFieldCount; i++)
                {
                    var fn = rdr.GetName(i); // tag or field name
                    if (acc == "text/html") // html headers have already been done
                        r.Append("<td>"+Format1(rdr[i])+"</td>");
                    else
                        r.Append(c + fmt[0] + fn  + fmt[1] +Format(rdr[i]));
                    if (acc == "application/xml") // xml need end tags
                        r.Append("</" + fn + ">");
                    c = fmt[6];
                }
                r.Append(fmt[7]);
            }
            return r + fmt[4];
        }
        /// <summary>
        /// This method allows for transactional operation: 
        /// the data is generally a sequence of SQL statements,
        /// to be performed in a transaction.
        /// We also allow queries: in which case the result sets (Json arrays) are simply concatenated.
        /// The current implementation assumes everyone agrees to use SQL as in ISO 9075:2016 ;)
        /// </summary>
        /// <param name="data"></param>
        /// <returns></returns>
        string Post1(string data)
        {
            var tr = conn.BeginTransaction(System.Data.IsolationLevel.Serializable);
            var ret = "";
            foreach (var s in data.Split(';'))
            {
                var cmd = conn.CreateCommand();
                cmd.Transaction = tr;
                var cm = s.Trim();
                if (cm == "")
                    continue;
                cmd.CommandText = cm;
                cm = cm.ToUpper();
                Console.WriteLine(cm);
                if (cm.StartsWith("SELECT") || cm.StartsWith("TABLE"))
                {
                    var rdr = cmd.ExecuteReader();
                    var docar = new DocArray();
                    while (rdr.Read())
                    {
                        var doc = new Document();
                        for (var i = 0; i < rdr.VisibleFieldCount; i++)
                            doc.fields.Add(new KeyValuePair<string, object>(rdr.GetName(i), rdr.GetValue(i)));
                        docar.items.Add(doc);
                    }
                    rdr.Close();
                    ret += docar.ToString();
                }
                else
                    cmd.ExecuteNonQuery();
            }
            tr.Commit();
            if (ret == "")
                ret = "OK";
            return ret;
        }
        string Post2(string tb,DocArray d)
        {
            var cls = new StringBuilder("insert into " + tb + "(");
            var vls = new StringBuilder(") values ");
            var cm = "";
            foreach (Document e in d.items)
            {
                vls.Append(cm);
                vls.Append("(");
                Document.Format(e, (cm == "") ? cls : null, vls);
                vls.Append(")");
                cm = ",";
            }
            var cmd = conn.CreateCommand();
            cmd.CommandText = cls.ToString() + vls.ToString();
            cmd.ExecuteNonQuery();
            return "OK";
        }
        string Put(string tb,string wh,Document d)
        {
            var cmt = "update " + tb + " set ";
            var cm = "";
            foreach (var e in d.fields)
            {
                cmt += cm + e.Key + "=" + Format(e.Value);
                cm = ",";
            }
            var cmd = conn.CreateCommand();
            cmd.CommandText = cmt+" where "+wh;
            cmd.ExecuteNonQuery();
            return "OK";
        }
        string Delete(string tb,string wh)
        {
            var cmd = conn.CreateCommand();
            cmd.CommandText = "delete from " + tb;
            if (wh != null)
                cmd.CommandText += " where " + wh;
            return "" + cmd.ExecuteNonQuery() + " rows deleted";
        }
        string Format(object ob)
        {
            if (ob is DBNull)
                return "null";
            if (ob is DateTime)
                return ((DateTime)ob).ToString("yyyy-MM-dd'T'HH:mm'Z'"); // ISO8601
            if (ob is string)
                return "'" + ob.ToString().Replace("'","''") + "'";
            if (ob is Document)
            {
                var sb = new StringBuilder();
                var cm = "";
                var d = (Document)ob;
                foreach(var x in d.fields)
                {
                    sb.Append(cm); cm = ",";
                    Document.Field(x, sb);
                }
                return sb.ToString();
            }
            return ob.ToString();
        }
        string Format1(object ob)
        {
            if (ob is DBNull)
                return "null";
            if (ob is DateTime)
                return ((DateTime)ob).ToString("yyyy-MM-dd'T'HH:mm'Z'"); // ISO8601
            return ob.ToString();
        }
        string Receive()
        {
            var r = context.Request;
            if (!r.HasEntityBody)
                return "No data??";
            var rdr = new StreamReader(r.InputStream, r.ContentEncoding);
            var rs = rdr.ReadToEnd();
            r.InputStream.Close();
            rdr.Close();
            return rs;
        }
        void SendCORS()
        {
            var h = context.Request.Headers["Access-Control-Request-Headers"];
            var c = context.Response;
            c.StatusCode = 200;
            c.StatusDescription = "OK";
            c.AddHeader("Acess-Control-Allow-Origin", origin);
            c.AddHeader("Acess-Control-Allow-Headers", h);
            c.AddHeader("Acess-Control-Allow-Methods", "GET, POST, PUT, DELETE, OPTIONS");
            c.AddHeader("Acess-Control-Max-Age", "86400");
            c.ContentLength64 = 0;
            c.OutputStream.Close();
        }
        void Send(int status,string mess)
        {
            var b = Encoding.UTF8.GetBytes(mess);
            var c = context.Response;
            c.StatusCode = status;
            c.StatusDescription = (status == 200) ? "OK" : "ERROR";
            if (origin!=null)
                c.AddHeader("Acess-Control-Allow-Origin", origin);
            c.AddHeader("Cache-control", "no-store");
            c.AddHeader("Expires", "-1");
            c.AddHeader("Pragma", "no-cache");
            if (mess != null && mess.Length > 0 && mess[0] == '[')
                c.AddHeader("Content-Type", "application/json");
            c.ContentLength64 = b.Length;
            var st = c.OutputStream;
            st.Write(b, 0, b.Length);
            st.Close();
        }
    }
}
